tests/ui/shared_code_in_if_blocks.rs

```diff
@@ -0,0 +1,198 @@
+    if x == 10 {
+        println!("x is a value");
+        println!("x is 10");
+    } else if x == 1 {
+        println!("x is 1");
+    } else {
+        println!("x is a value");
+    }
+
+    // Shared code in the middle
+    let _ = if x == 7 {
+        let mut z = 3;
+        println!("x is a value");
+        z *= 2;
+        z
+    } else if x == 100 {
+        let mut y = 1;
+        println!("x is a value");
+        y += 9;
+        y
+    } else {
+        let x = 2;
+        println!("x is a value");
+        x
+    };
+
+    // Returns value with the same name
+    let _ = if x == 9 {
+        let x = 9;
+        println!("--");
+        x
+    } else {
+        let x = 18;
+        println!("^^");
+        x
+    };
+
+    for index in 0..2 {
+        // Okay because we don't have an unconditional else case
+        if index == 6 {
+            println!("Six");
+            continue;
+        } else if index == 9 {
+            println!("Nine");
+            continue;
+        }
+
+        println!("---");
+    }
+}
+
+fn shared_code_at_start() {
+    let x = 17;
+
+    // First statement is the same
+    let _ = if x == 1 {
+        // I'm a comment that shouldn't matter for the lint
+        println!("Hello World");
+
+        4
+    } else {
+        // I'm another comment,
+        // Nice to see you here in the tests
+        println!("Hello World");
+
+        16
+    };
+
+    // First statement is the same
+    if x == 19 {
+        println!("Hello World");
+        let mut y = 9;
+        y += 1;
+        let _z = y;
+    } else {
+        println!("Hello World");
+    };
+
+    if x == 8 {
+        println!("Hello World");
+        println!("How are you today?");
+        println!("From: Test");
+        let _x = 10;
+        let _ = true;
+    } else if x != 2 {
+        println!("Hello World");
+        println!("How are you today?");
+        println!("From: Test");
+        let _x = 10;
+        let _ = false;
+    } else {
+        println!("Hello World");
+        println!("How are you today?");
+        println!("From: Test");
+        let _x = 10;
+    }
+}
+
+fn shared_code_at_end() {
+    let x = 10;
+
+    // Same code at the end
+    if x == 8 {
+        let _ = true;
+        println!("Hello World");
+    } else if x != 2 {
+        let _ = false;
+        println!("Hello World");
+    } else {
+        println!("Hello World");
+    }
+
+    let _z = if x == 8 {
+        println!("Branch 1");
+        let z = 10;
+        foo!();
+        z
+    } else {
+        println!("Branch 2");
+        let z = 10;
+        foo!();
+        z
+    };
+
+    // Lint at start and end
+    let _ = if x == 1 {
+        println!("I'm the same as my brother branch");
+        let _a = 99;
+        println!("End of block");
+        false
+    } else {
+        println!("I'm the same as my brother branch");
+        let _b = 17;
+        println!("End of block");
+        false
+    };
+}
```

Comment:
I have a feeling that many tests are missing. Mainly tests where the returned value is present somewhere in the code that should be moved out. I think we should just not lint if the return value is used in the end stmts, but can lint if it is only used in the start stmts.

Reply:
I think that even these cases should be included in the lint since it can still be code duplication. My previous [comment](https://github.com/rust-lang/rust-clippy/pull/6463#discussion_r549440565) shows a rewrite where the return value is not a part of the linted code.

The main reason why I'm not in favor of your suggestion is the implementation. This may enabled me to reuse the provided `SpanlessEq` implementation with very few changes. Adding the simple rule that the return value needs to be in the moved code would in my opinion add a lot of edge cases and limit the use of `SpanlessEq` since it wasn't designed for value tracking. It can quickly end up in a lot of complex code which is difficult to maintain for little benefit. 

That's my opinion with my current knowledge but I might have overlooked something. I'm still fairly new to clippy and I still learn something every time I work on it 🙃.

I've implemented your other suggestions. Thank you for those, they where definitely helpful 👍. I'll wait on an answer on this comment before adding a bunch of tests 🙃

Reply:
I think rewriting it like you suggested above makes it way harder to read and we shouldn't suggest that. Also this is not always possible. For example if `z` is modified, but not returned. With my suggested change, I don't think you have to modify existing code, but rather just add a visitor and an additional check.

In general, I don't think a lint just saying "this is duplicated" without an easy suggestion how to improve it, is a good lint. We shouldn't say: "this is wrong, have fun rewriting it" without giving guidance on how to do it.

Reply:
Jup, the rewrite is definitely not pretty 😅. 

The lint currently only suggests moving the code out of the block. Making a correct concrete suggestion seamed difficult as the lint didn't make sure that the moved values will also be available after the load. This should now be possible with your suggested change to make sure that the return value is included in the moved code or at least available in the outer scope. It will need some fine-tuning and probably _some_ testing but should be possible 🙃.